src/server/chromium/crNetworkManager.ts

```diff
@@ -542,3 +563,133 @@ const errorReasons: { [reason: string]: Protocol.Network.ErrorReason } = {
   'timedout': 'TimedOut',
   'failed': 'Failed',
 };
+
+type RequestInfo = {
+  requestId: string,
+  responseReceived: boolean,
+  responseReceivedExtraInfo: Protocol.Network.responseReceivedExtraInfoPayload[],
+  responses: network.Response[],
+  loadingFinished?: Protocol.Network.loadingFinishedPayload,
+  loadingFailed?: Protocol.Network.loadingFailedPayload
+};
+
+//
+// This class aligns responses with response headers from extra info:
+//   - Network.requestWillBeSent, Network.responseReceived, Network.loadingFinished/loadingFailed are
+//     dispatched using one channel.
+//   - Network.requestWillBeSentExtraInfo and Network.responseReceivedExtraInfo are dispatches on
+//     another channel. Those channels are not associated, so events come in random order.
+//
+// This class will associate responses with the new headers. These extra info headers will become
+// available to client reliably upon requestfinished event only.
+//
+class ResponseExtraInfoTracker {
+  private _requests = new Map<string, RequestInfo>();
+
+  constructor() {
+  }
+
+  requestWillBeSent(event: Protocol.Network.requestWillBeSentPayload) {
+    const info = this._requests.get(event.requestId);
+    if (info) {
+      // This is redirect.
+      if (!event.redirectResponse?.connectionId) {
+        // Bail out early for responses without connectionId.
+        // They won't be getting extra info, so there is nothing to track.
+        this._stopTracking(event.requestId);
+        return;
+      }
+      info.responseReceived = true;
+    } else {
+      this._requests.set(event.requestId, {
+        requestId: event.requestId,
+        responseReceived: false,
+        responseReceivedExtraInfo: [],
+        responses: [],
+      });
+    }
+  }
+
+  responseReceived(event: Protocol.Network.responseReceivedPayload) {
+    const info = this._requests.get(event.requestId);
+    if (!info)
+      return;
+
+    // Bail out early for responses without connectionId.
```

Comment:
Why don't we do this in processResponse?

Reply:
To follow a contract cleanly. If we've ever seen a CDP response w/o connectionId, we aren't interested in the services of this tracker.